### PR TITLE
Update Calico to 3.1

### LINKED ISSUE #1621: k8s 1.8 template generated by acs-engine no longer working
**Is this a request for help?**:

No; bug report
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

ISSUE
---

**What version of acs-engine?**:

#5b57309 was used to generate the template. 
---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**

k8s 1.8

**What happened**:

One week ago I generated the following template with acs-engine. k8s deployed fine from the template 3-4 times in a row. The template is no longer provisioning a working k8s 1.8 cluster. 

https://gist.githubusercontent.com/zimmertr/e898f95077181f3a089cd5896f2f95aa/raw/6053d6edb879264bad2b8c59758301ee125a3154/gistfile1.txt

`Oct 19 15:18:40 k8s-master-48084675-0 docker[1641]: W1019 15:18:40.700140    1838 status_manager.go:431] Failed to get status for pod "calico-node-r9n9g_kube-system(c796464e-b46b-11e7-8b57-000d3a220cf3)": client: etcd cluster is unavailable or misconfigured; error #0: client: etcd member http://127.0.0.1:2379 has no leader`

```
user@k8s-master-48084675-0:~$ etcdctl member list
33b0bb17e22c066a: name=k8s-master-48084675-0 peerURLs=http://172.17.38.30:2380 clientURLs=http://172.17.38.30:2379
7f03ea887d8488f0: name=k8s-master-48084675-2 peerURLs=http://172.17.38.32:2380 clientURLs=http://172.17.38.32:2379
a4058da44ae0514c: name=k8s-master-48084675-1 peerURLs=http://172.17.38.31:2380 clientURLs=http://172.17.38.31:2379
```

**How to reproduce it** (as minimally and precisely as possible):

Deploy the template listed above. The template is meant to be deployed on top of this network schema: 

https://gist.githubusercontent.com/zimmertr/d03d04ae5b1325af25095aa6aa5b12db/raw/3b254d65b19a06d769febd4acfcf4dfce1232090/gistfile1.txt

**Anything else we need to know**:

This template was working 100% successfully before. It is randomly no longer working. 

Calico released a new version (2.6.2) 3 days ago which may be the culprit. But they claim no changes were made to k8s in the release notes: https://github.com/projectcalico/calico/releases/tag/v2.6.2

However, this pull request addressed a very similar Calico issue when k8s 1.8 first came out: https://github.com/Azure/acs-engine/pull/1511

It's also possible this commit broke it as it effected `etcd` recently as well: https://github.com/Azure/acs-engine/pull/1564

### LINKED ISSUE #2202: Multi-Master CNI config Not Present On All Masters
**Is this a request for help?**:

---No

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

---Issue

**What version of acs-engine?**:

---0.12.5

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes

**What happened**: When deploying a template that contains multiple masters the calico CNI config doesen't get written to all master nodes. Ran through it three different times and each time only one master picked up the calico CNI config. Copying the net.d/ calico config from healthy master fixed the issue

**What you expected to happen**: Healthy cluster with all CNI configs present

**How to reproduce it** (as minimally and precisely as possible):
Run config with multiple masters like below 
{
    "apiVersion": "vlabs",
    "properties": {
        "orchestratorProfile": {
            "orchestratorType": "Kubernetes",
            "orchestratorRelease": "1.9",
            "kubernetesConfig": {
                "networkPolicy": "calico",
                "enableDataEncryptionAtRest": true
            }
        },
        "masterProfile": {
            "count": 3,
            "dnsPrefix": "xx-xxxxxxxx",
            "vmSize": "Standard_DS2_v2_Promo"
        },

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by dtzar**
**What this PR does / why we need it**:
Updates Calico from 2.6 to the latest 3.0 which enables [additional functionality](https://www.projectcalico.org/releasethecalico/). 

**Which issue this PR fixes** 
This lays the foundation for the requested IPAM support via these issues 
https://github.com/Azure/acs-engine/issues/2227
https://github.com/Azure/acs-engine/issues/1930

Potentially fixes #1621
fixes #2202

**If applicable**:
- [x] documentation
- [ ] network policy e2e test(s)
- [x] tested backward compatibility (ie. deploy with the previous version, upgrade with this branch)

```release-note
Updates Calico from 2.6 to the latest 3.0
```

### CODE REVIEW

**Comment by tmjd** [comment]
LGTM from a Calico perspective for the changes to parts/k8s/addons/kubernetesmasteraddons-calico-daemonset.yaml

**Comment by jackfrancis** [comment]
@dtzar is this PR ready for further testing/merging when the merge conflict is fixed?

**Comment by dtzar** [comment]
@jackfrancis - Resolved the merge conflict.  This is ready for further testing/merging.  Would be great to let Circle CI do all the existing e2e tests as they are now.

Sorry, I just haven't had a chance to work on creating a new e2e test for Calico.

**Comment by jackfrancis** [comment]
We'll do the standard drill, and then validate that `examples/networkpolicy/kubernetes-calico.json` can build a working k8s cluster.

As you mention, we don't have calico-specific E2E validations yet, do you want to wait for those to merge? What's your manual testing telling you at this point?

File: examples/networkpolicy/kubernetes-calico.json [hunk 1]
```diff
@@ -10,19 +10,13 @@
     "masterProfile": {
       "count": 1,
       "dnsPrefix": "",
-      "vmSize": "Standard_D2_v2"
+      "vmSize": "Standard_DS2_v2"
```
Comment (jackfrancis) [inline examples/networkpolicy/kubernetes-calico.json]: Any reason for the change to the VM SKU in the example api model?

Comment (dtzar) [reply examples/networkpolicy/kubernetes-calico.json]: Yes, this uses the faster premium storage based VMs.

**Comment by jackfrancis** [comment]
@dtzar How are you getting past the lack of CNI implementation in your testing?

```
azureuser@k8s-master-19545415-0:~$ ls -la /etc/cni/net.d
ls: cannot access '/etc/cni/net.d': No such file or directory
```

I'm getting this in kubelet logs:

```
cni.go:196] Unable to update cni config: No networks found in /etc/cni/net.d
```

tldr; current calico policy sets kubelet runtime config to `--network-policy=cni`. How does this work in this implementation of calico?

File: parts/k8s/addons/kubernetesmasteraddons-calico-daemonset.yaml [hunk 2]
```diff
@@ -265,14 +274,14 @@ spec:
...
         k8s-app: calico-node
-      annotations:
-        # This, along with the CriticalAddonsOnly toleration below,
-        # marks the pod as a critical add-on, ensuring it gets
-        # priority scheduling and that its resources are reserved
-        # if it ever gets evicted. 
-        # Deprecated in 1.10, Removed in 1.11. kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods 
-        scheduler.alpha.kubernetes.io/critical-pod: ''
-    spec:
+        annotations:
```
Comment (dtzar) [inline parts/k8s/addons/kubernetesmasteraddons-calico-daemonset.yaml]: I gave myself a problem here.  Fixing now.

**Comment by dtzar** [comment]
After my latest fix, I validated manually that calico worked properly with the advanced policy demo on a 1.8 cluster.  Something went wrong on 1.10 - looking into this now.

**Comment by dtzar** [comment]
Wasn't able to fix the 1.10 issue, but did check again that 1.9 works also.
